lib/mongodb.js

```diff
@@ -2023,28 +2038,49 @@ function visitAllProperties(data, modelCtorOrDef, visitor) {
         visitAllProperties(value, def.type.definition, visitor);
       }
     } else {
-      visitor(value, def, (newValue) => { data[p] = newValue; });
+      visitor(model, value, def, (newValue) => { data[p] = newValue; });
     }
     continue;
   }
 }
 
 /**
- *
- * @param {*} propValue Property value to coerce into a Decimal128 value
- * @param {*} propDef Property definition to check if property is MongoDB
- * Decimal128 type
+ * @param {*} model Model containing the property
+ * @param {*} propValue Property value to coerce into special types supported by the connector
+ * @param {*} propDef Property definition to check if property is for MongoDB
  */
-function coerceDecimalProperty(propValue, propDef, setValue) {
-  let updatedValue;
-  if (hasDataType('decimal128', propDef)) {
-    if (Array.isArray(propValue)) {
-      updatedValue = propValue.map(val => Decimal128.fromString(val));
-      return setValue(updatedValue);
-    } else {
-      updatedValue = Decimal128.fromString(propValue);
-      return setValue(updatedValue);
+function coerceProperties(model, propValue, propDef, setValue) {
+  let coercedValue;
+  // Process only mongo-specific data types
+  if (propDef && propDef.mongodb && propDef.mongodb.dataType) {
+    const dataType = propDef.mongodb.dataType;
+    // Data types specified as string values
+    // Yaapa: should we implement all type specifications using objects instead of strings?
+    if (typeof dataType === 'string') {
+      if (hasDataType('decimal128', propDef)) {
+        if (Array.isArray(propValue)) {
+          coercedValue = propValue.map(val => Decimal128.fromString(val));
+          return setValue(coercedValue);
+        } else {
+          coercedValue = Decimal128.fromString(propValue);
+          return setValue(coercedValue);
+        }
+      }
+    // ObjectID data type
+    } else if (dataType === ObjectID) {
+      if (!model.settings.strictObjectIDCoercion) {
+        coercedValue = ObjectID(propValue);
+        setValue(coercedValue);
+      }
     }
+  } else {
+    // Apply strictObjectIDCoercion setting to ObjectID-like string values
+    if (typeof propValue === 'string' && !model.settings.strictObjectIDCoercion && propValue.match(/^[0-9a-fA-F]{24}$/)) {
```

Comment:
Why are we checking `model.settings.strictObjectIDCoercion` again? The rule to decide whether a string value should be coerced to ObjectID is more complex than a RegExp check, that's why we have `isObjectIDProperty`.